### PR TITLE
Improve error reports here and there

### PR DESCRIPTION
**Pull request by SebastianMestre**
I wanted errors to look nicer for the next alpha. This is a step towards that.

### CODE REVIEW

**Comment by SebastianMestre** [comment]
This was kinda fun ngl

File: src/typesystem.cpp [hunk 1]
```diff
@@ -59,7 +73,9 @@ TypeSystemCore::TypeSystemCore() {
 			b_data.argument_count = new_argument_count;
 
 		} else {
-			assert(0 and "unifying two different known type functions");
+			Log::fatal(
+			    "Deduced two different type functions to be equal (with IDs " +
+			    std::to_string(fa) + " and " + std::to_string(fb) + ")");
```
Comment (lushoBarlett) [inline src/typesystem.cpp:78]: Could use `FatalStream`

File: src/typecheck.cpp [hunk 2]
```diff
@@ -293,9 +293,10 @@ void print_information(TypedAST::Declaration* ast, TypeChecker& tc) {
 #if DEBUG
 	auto poly = ast->m_decl_type;
 	auto& poly_data = tc.m_core.poly_data[poly];
-	std::cerr << "@@ Type of local variable " << ast->identifier_text() << '\n';
-	std::cerr << "@@ Has " << poly_data.vars.size() << " variables\n";
-	std::cerr << "@@ It is equal to:\n";
+	Log::info(
+	    "Type of local variable '" + ast->identifier_text().str() + "' has " +
+	    std::to_string(poly_data.vars.size()) + " type variables");
```
Comment (lushoBarlett) [inline src/typecheck.cpp:298]: A general purpose logging stream will be the next iteration of the logging system probably.

File: src/match_identifiers.cpp [hunk 3]
```diff
@@ -307,9 +323,7 @@ namespace TypeChecker {
 
 #undef DO_NOTHING
 #undef DISPATCH
-	std::cerr << "INTERNAL ERROR: UNHANDLED CASE IN " << __PRETTY_FUNCTION__
-	          << ": " << typed_ast_string[(int)ast->type()] << '\n';
-	assert(0);
+	Log::fatal(std::string("(internal) Unhandled case in match_identifiers '") + typed_ast_string[int(ast->type())] + "'");
```
Comment (lushoBarlett) [inline src/match_identifiers.cpp:326]: `FatalStream`

File: src/match_identifiers.cpp [hunk 4]
```diff
@@ -41,7 +54,7 @@ namespace TypeChecker {
 	if (!declaration) {
 		// TODO: clean up how we build error reports
 		return {
-		    "ERROR @ line " + std::to_string(ast->m_token->m_line0 + 1) +
+		    "at line " + std::to_string(ast->m_token->m_line0 + 1) +
 		    " : accessed undeclared identifier '" + ast->text().str() + "'"};
```
Comment (lushoBarlett) [inline src/match_identifiers.cpp:58]: Perhaps allowing `sstream` in our error creation will be beneficial for our next iterations.

File: src/log/fatal_stream.cpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+#define INSTANCE_SHIFT_OPERATOR(type)                                          \
+	FatalStream& FatalStream::operator<<(type in) {                            \
+		std::cerr << in;                                                       \
+		return *this;                                                          \
+	}
+
+INSTANCE_SHIFT_OPERATOR(std::string const&)
+INSTANCE_SHIFT_OPERATOR(int)
+INSTANCE_SHIFT_OPERATOR(float)
+INSTANCE_SHIFT_OPERATOR(char)
```
Comment (lushoBarlett) [inline src/log/fatal_stream.cpp:16]: Cute.

File: src/compute_offsets.cpp [hunk 6]
```diff
@@ -208,9 +207,9 @@ void compute_offsets(TypedAST::TypedAST* ast, int frame_offset) {
 
 #undef DO_NOTHING
 #undef DISPATCH
-	std::cerr << "INTERNAL ERROR: UNHANDLED CASE IN " << __PRETTY_FUNCTION__
-	          << ": " << typed_ast_string[(int)ast->type()] << '\n';
-	assert(0);
+	Log::fatal(
+	    std::string("(internal) Unhandled case in compute_offsets (") +
+	    typed_ast_string[(int)ast->type()] + ")");
```
Comment (lushoBarlett) [inline src/compute_offsets.cpp:212]: `FatalStream`

**Review by lushoBarlett — APPROVED** [review]
FatalSteam is a nice idea. I like how you instanced functions with a macro, very _ruby-esque_.

File: src/log/fatal_stream.cpp (hunk 5 above)
Comment (SebastianMestre) [reply src/log/fatal_stream.cpp:16]: yeah pretty cute innit

File: src/typecheck.cpp (hunk 2 above)
Comment (SebastianMestre) [reply src/typecheck.cpp:298]: Yeah, we probably want a `BasicLogStream` or something that the other ones inherit from. (future, lol)

File: src/match_identifiers.cpp (hunk 4 above)
Comment (SebastianMestre) [reply src/match_identifiers.cpp:58]: I would like to integrate the `Log` stuff with `ErrorReport` somehow... I dunno

**Comment by SebastianMestre** [comment]
> FatalSteam is a nice idea. I like how you instanced functions with a macro, very _ruby-esque_.

lol. I could've used a template, but this just seemed simpler y'no

**Pull request merged**